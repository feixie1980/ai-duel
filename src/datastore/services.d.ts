export type ServiceType = 'openai' | 'deepseek';

export type ModelType = 'gpt-4o-mini';
