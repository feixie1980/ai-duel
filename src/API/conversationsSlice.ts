import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { v4 as uuidv4 } from 'uuid';

type Persona = 'user' | 'llm';
type MessageStatus = 'pending' | 'completed' | 'error';

const STORAGE_KEY = 'conversations';

const loadFromStorage = (): Conversation[] => {
  try {
    const stored = localStorage.getItem(STORAGE_KEY);
    if (!stored) return [];
    return JSON.parse(stored);
  } catch (error) {
    console.error('Failed to load conversations from storage:', error);
    return [];
  }
};

const saveToStorage = (conversations: Conversation[]): void => {
  try {
    localStorage.setItem(STORAGE_KEY, JSON.stringify(conversations));
  } catch (error) {
    console.error('Failed to save conversations to storage:', error);
  }
};

export interface ChatMessage {
  id: string;
  content: string;
  from: Persona;
  status: MessageStatus;
}

export interface Conversation {
  id: string;
  timestamp?: number;
  messages: ChatMessage[];
}

export interface ConversationsState {
  conversations: Conversation[];
  currentConversationId: string | undefined;
}

const initialState: ConversationsState = {
  conversations: loadFromStorage(),
  currentConversationId: undefined,
};

// slice
const conversationSlice = createSlice({
  name: 'conversations',
  initialState,
  reducers: {
    initEmptyConversation: (state) => {
      state.currentConversationId = undefined;
    },

    addConversation: (
      state,
      action: PayloadAction<{
        initialMessages: ChatMessage[];
      }>
    ) => {
      const newConversation: Conversation = {
        id: uuidv4(),
        timestamp: new Date().getTime(),
        messages: action.payload.initialMessages,
      };
      state.conversations.push(newConversation);
      state.currentConversationId = newConversation.id;

      saveToStorage(state.conversations);
    },

    addMessage: (
      state,
      action: PayloadAction<{
        conversationId: string;
        content: string;
        from: Persona;
      }>
    ) => {
      const { conversationId, content, from } = action.payload;
      if (!conversationId) {
        return;
      }
      const conversation = state.conversations.find(
        (conv) => conv.id === conversationId
      );
      if (!conversation) {
        throw `conversation with id ${conversationId} not found!`;
      }

      const noPending = !conversation.messages.some(
        (m) => m.status === 'pending'
      );

      if (!noPending) {
        // if there is a pending response from LLM provider, do not add the new message
        return;
      }

      conversation.messages.push({
        id: uuidv4(),
        content,
        from,
        status: from === 'user' ? 'completed' : 'pending',
      });

      saveToStorage(state.conversations);
    },

    updateMessage: (
      state,
      action: PayloadAction<{
        conversationId: string;
        messageId: string;
        content: string;
        status: MessageStatus;
      }>
    ) => {
      const { conversationId, messageId, content, status } = action.payload;
      if (!conversationId) {
        return;
      }
      const conversation = state.conversations.find(
        (conv) => conv.id === conversationId
      );
      if (!conversation) {
        throw `conversation with id ${conversationId} not found!`;
      }

      const message = conversation.messages.find((m) => m.id === messageId);
      if (!message) {
        throw `message with id ${messageId} not found in conversation with id ${conversationId}!`;
      }
      message.status = status;
      message.content = content;

      saveToStorage(state.conversations);
    },
  },
});

// Actions
export const { initEmptyConversation, addConversation, updateMessage } =
  conversationSlice.actions;

// Selectors
export const selectCoversations = (state: ConversationsState) =>
  state.conversations;

export const selectCurrentConversation = (state: ConversationsState) => {
  const { conversations, currentConversationId } = state;
  if (!currentConversationId) {
    return null;
  }
  return conversations.find((c) => c.id === currentConversationId);
};

export default conversationSlice.reducer;
