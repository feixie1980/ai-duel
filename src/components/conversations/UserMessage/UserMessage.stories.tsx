import type { Meta, StoryObj } from '@storybook/react';
import UserMessage from './UserMessage';

const meta = {
  title: 'Components/conversations/UserMessage',
  component: UserMessage,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
} satisfies Meta<typeof UserMessage>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {
    message: {
      id: '1',
      content: 'Hello World',
      from: 'llm',
      status: 'completed',
    },
  },
};
