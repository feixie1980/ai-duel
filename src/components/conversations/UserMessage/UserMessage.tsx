import { Box } from '@radix-ui/themes';
import styled from 'styled-components';
import { UserMessageProps } from './UserMessage.types';

const UserMessageBox = styled(Box)`
  align-self: flex-end;
  background-color: var(--gray-2);
  padding: 12px;
  border-radius: 8px;
`;

export default function UserMessage(props: UserMessageProps) {
  const { message } = props;
  return <UserMessageBox>{message.content}</UserMessageBox>;
}
