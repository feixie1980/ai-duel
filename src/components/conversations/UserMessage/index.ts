export { default as UserMessage } from './UserMessage'; 
export type { UserMessageProps } from './UserMessage.types';