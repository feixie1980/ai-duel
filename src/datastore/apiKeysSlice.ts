import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import CryptoJS from 'crypto-js';
import { ServiceType } from './services';

// Types
interface APIKey {
  service: ServiceType;
  encryptedKey: string;
}

interface APIKeysState {
  keys: APIKey[];
  loading: boolean;
  error: string | null;
}

// Encryption utilities
const STORAGE_KEY = 'api_keys';
const getEncryptionKey = () => {
  const key = import.meta.env.VITE_ENCRYPTION_KEY;
  if (!key) {
    throw new Error('Encryption key not found in environment variables');
  }
  return key;
};

const encryptKey = (plainText: string): string => {
  return CryptoJS.AES.encrypt(plainText, getEncryptionKey()).toString();
};

const decryptKey = (cipherText: string): string => {
  const bytes = CryptoJS.AES.decrypt(cipherText, getEncryptionKey());
  return bytes.toString(CryptoJS.enc.Utf8);
};

// LocalStorage utilities

/* 
  In real production app, api keys should not be stored on client because of
  security risk. We are using local storage just for convenience and testing
  purpose, becuase I don't want to write server yet~
*/

// TODO: change this to session storage, a tiny bit more secure.
const loadFromStorage = (): APIKey[] => {
  try {
    const stored = localStorage.getItem(STORAGE_KEY);
    if (!stored) return [];
    return JSON.parse(stored);
  } catch (error) {
    console.error('Failed to load API keys from storage:', error);
    return [];
  }
};

const saveToStorage = (keys: APIKey[]): void => {
  try {
    localStorage.setItem(STORAGE_KEY, JSON.stringify(keys));
  } catch (error) {
    console.error('Failed to save API keys to storage:', error);
  }
};

// Initial state
const initialState: APIKeysState = {
  keys: loadFromStorage(),
  loading: false,
  error: null,
};

// Slice
const apiKeysSlice = createSlice({
  name: 'apiKeys',
  initialState,
  reducers: {
    addAPIKey: (
      state,
      action: PayloadAction<{
        service: ServiceType;
        key: string;
      }>
    ) => {
      const { service, key } = action.payload;
      const encryptedKey = encryptKey(key);

      // Remove existing key for the same service if it exists
      state.keys = state.keys.filter((k) => k.service !== service);

      // Add new key
      state.keys.push({ service, encryptedKey });
      saveToStorage(state.keys);
    },

    removeAPIKey: (state, action: PayloadAction<ServiceType>) => {
      state.keys = state.keys.filter((key) => key.service !== action.payload);
      saveToStorage(state.keys);
    },

    updateAPIKey: (
      state,
      action: PayloadAction<{
        service: ServiceType;
        key: string;
      }>
    ) => {
      const { service, key } = action.payload;
      const encryptedKey = encryptKey(key);

      const index = state.keys.findIndex((k) => k.service === service);
      if (index !== -1) {
        state.keys[index] = { service, encryptedKey };
        saveToStorage(state.keys);
      }
    },

    setError: (state, action: PayloadAction<string | null>) => {
      state.error = action.payload;
    },
  },
});

// Actions
export const { addAPIKey, removeAPIKey, updateAPIKey, setError } =
  apiKeysSlice.actions;

// Selectors
export const selectAPIKeys = (state: { apiKeys: APIKeysState }) =>
  state.apiKeys.keys;

export const selectAPIKeyByService =
  (service: ServiceType) => (state: { apiKeys: APIKeysState }) => {
    const key = state.apiKeys.keys.find((k) => k.service === service);
    if (!key) return null;

    try {
      const decryptedKey = decryptKey(key.encryptedKey);
      return decryptedKey;
    } catch (error) {
      console.error('Failed to decrypt API key:', error);
      return null;
    }
  };

export const selectError = (state: { apiKeys: APIKeysState }) =>
  state.apiKeys.error;

export default apiKeysSlice.reducer;
