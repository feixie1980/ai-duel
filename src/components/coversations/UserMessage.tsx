import { Box } from '@radix-ui/themes';
import { ChatMessage } from '../../API/conversationsSlice';
import styled from 'styled-components';

const UserMessageBox = styled(Box)`
  align-self: flex-end;
  background-color: var(--gray-2);
  padding: 12px;
  border-radius: 8px;
`;

export function UserMessage({ message }: { message: ChatMessage }) {
  return <UserMessageBox>{message.content}</UserMessageBox>;
}
