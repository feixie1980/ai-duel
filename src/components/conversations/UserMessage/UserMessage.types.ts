import { ChatMessage } from '../../../datastore/conversationsSlice';

export interface UserMessageProps {
  /**
   * Custom class name for the container of the component.
   */
  className?: string;

  message: ChatMessage;
}
