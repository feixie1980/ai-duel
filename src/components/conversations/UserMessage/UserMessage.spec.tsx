import { describe, it, expect } from 'vitest';
import { render } from '@testing-library/react';
import UserMessage from './UserMessage';
import { ChatMessage } from '../../../datastore/conversationsSlice';

describe('UserMessage', () => {
  it('renders correctly', () => {
    const message: ChatMessage = {
      id: '1',
      content: 'Hello World',
      from: 'user',
      status: 'completed',
    };
    const { container } = render(<UserMessage message={message} />);
    expect(container).toMatchSnapshot();
  });
});
