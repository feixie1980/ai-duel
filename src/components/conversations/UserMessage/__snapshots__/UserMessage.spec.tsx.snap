// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UserMessage > renders correctly 1`] = `
<div>
  <div
    class="rt-Box UserMessage__UserMessageBox-sc-4b9563b2-0 imGPAl"
  >
    Hello World
  </div>
</div>
`;
